accepted/cranelift-assembler.md

```diff
@@ -0,0 +1,368 @@
+  option would be to extend ISLE with macro-like functionality. Feedback is welcome on this point!
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+> What other designs have been considered and what is the rationale for choosing the one proposed?
+
+### Why a new assembler?
+[Why a new assembler?]: #why-a-new-assembler
+
+One good argument against this proposal is that it reinvents the wheel. Surely a new assembler need
+not be written when so many battle-tested assemblers already exist?
+
+This proposal does not advocate creating a new assembler, though. It would appear so because new
+crates appear, `cranelift-assembler-*`, but truly this assembler was already in existence inside
+`cranelift-codegen`. By separating what was already there, this proposal promises better testing,
+better verification integration, possible reuse, etc.
+
+In designing the prototype, I (@abrown) evaluated whether an already-existing assembler could be
+slotted in here instead of moving the existing `cranelift-codegen` logic. My opinion is that the
+difficult part is the [Integration] bits and that using an existing assembler (e.g., [XED]) does not
+help with that. In fact, having control over the generated code made it easier to adapt to
+Cranelift's peculiarities quickly.
+
+### Why not keep the status quo?
+[Why not keep the status quo]: #why-not-keep-the-status-quo
+
+The prototype is based on x64 and the irregularities of that ISA may make it ideal for this kind of
+project. Will other backends want to follow a similar path? It's unclear whether other backends want
+to (or should) follow the split made by this proposal.
+
+In favor of moving to this proposal, backends like aarch64 have also gained many `MInst` "category"
+variants; they may benefit from this kind of split. Arguing against, though: this could be work that
+other backend maintainers don't want to take on. This seems like a good topic for discussion &mdash;
+feel free to comment on this!
+
+One practical point from experience prototyping is that, while it seems clearly possible to generate
+assembler instructions for other ISAs, the details of how to emit, pretty-print, etc. could be quite
+different. This means the generation code is not a build-once-for-any-ISA affair. It may make sense
+to separately finish out the `cranelift-assembler-x64` work before trying to generalize to other
+ISAs.
+
+# Open questions
+[open-questions]: #open-questions
+
+> - What parts of the design do you expect to resolve through the RFC process before this gets
+>   merged?
+> - What parts of the design do you expect to resolve through implementation after the RFC is
+>   accepted?
+
+- How should we handle the additional verbosity that comes from materializing all the machine code
+  instructions?
+- Which backends want to migrate to this proposal? Should we plan to migrate all backends to this
+  proposal?
```

Comment:
I think ideally we want to move everything over towards something like whatever we end up at here. I don't think everything needs to use exactly the same machinery, but the motivation for x64 applies pretty much to every backend.

We also don't have to do the migration en masse, all at once, and before experimenting with just the x64 backend. We can do things in an incremental fashion and learn what works well and what doesn't as we go.